### ISSUE TITLE
TLS client-initiated renegotiations are supported unexpectedly

### ISSUE THREAD

**Issue by bratkartoffel** [proftpd/proftpd#1076]
### What I Did

~~The latest release does not handle various tls security related settings as defined by documentation.~~ (Update: https://github.com/proftpd/proftpd/issues/1076#issuecomment-689015778)
The latest release can be DOS'ed as client-initiated renegotiations are allowed.

### What I Expected/Wanted

Client initiated renegotiations should be disabled ~~and server cipher suite preffered order should be used.~~ (Update: https://github.com/proftpd/proftpd/issues/1076#issuecomment-689015778)

### ProFTPD Version and Configuration

proftpd 1.3.7a from alpine edge:
```
Compile-time Settings:
  Version: 1.3.7a (maint)
  Platform: LINUX [Linux 4.15.0-112-generic x86_64]
  Built: Sat Jul 25 2020 21:12:02 UTC
  Built With:
    configure  '--build=x86_64-alpine-linux-musl' '--host=x86_64-alpine-linux-musl' '--prefix=/usr' '--sysconfdir=/etc/proftpd' '--localstatedir=/run/proftpd' '--libexecdir=/usr/lib/proftpd' '--disable-static' '--enable-sendfile' '--enable-facl' '--enable-dso' '--enable-autoshadow' '--enable-ctrls' '--enable-ipv6' '--enable-pcre' '--disable-strip' '--disable-wtmp' '--disable-auth-pam' '--enable-openssl' '--with-shared=mod_unique_id:mod_site_misc:mod_load:mod_ban:mod_quotatab:mod_sql:mod_sql_mysql:mod_sql_postgres:mod_sql_sqlite:mod_sql_odbc:mod_dynmasq:mod_quotatab_sql:mod_ldap:mod_quotatab_ldap:mod_ratio:mod_tls:mod_rewrite:mod_radius:mod_wrap2:mod_wrap2_file:mod_wrap2_sql:mod_quotatab_file:mod_quotatab_radius:mod_facl:mod_ctrls_admin:mod_copy:mod_deflate:mod_ifversion:mod_geoip:mod_exec:mod_sftp:mod_sftp_sql:mod_shaper:mod_sql_passwd:mod_ifsession:mod_auth_otp:mod_tls_redis:mod_wrap2_redis:mod_redis:mod_memcache:mod_tls_memcache:mod_readme:mod_snmp:mod_tls_shmcache' 'build_alias=x86_64-alpine-linux-musl' 'host_alias=x86_64-alpine-linux-musl' 'CC=gcc' 'CFLAGS=-Os -fomit-frame-pointer' 'LDFLAGS=-Wl,--as-needed' 'CPPFLAGS=-Os -fomit-frame-pointer' 'CXX=g++' 'CXXFLAGS=-Os -fomit-frame-pointer'

  CFLAGS: -g2 -Os -fomit-frame-pointer -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration
  LDFLAGS: -L$(top_srcdir)/lib -L$(top_builddir)/lib -Wl,--as-needed -rdynamic  -L/usr/lib/ -L/usr/lib
  LIBS:  -lpcreposix -lpcre -lssl -lcrypto -lsupp -lresolv -lresolv -lcrypt  -pthread

  Files:
    Configuration File:
      /etc/proftpd/proftpd.conf
    Pid File:
      /run/proftpd/proftpd.pid
    Scoreboard File:
      /run/proftpd/proftpd.scoreboard
    Header Directory:
      /usr/include/proftpd
    Shared Module Directory:
      /usr/lib/proftpd

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    + Autoshadow support
    + Controls support
    - curses support
    - Developer support
    + DSO support
    + IPv6 support
    + Largefile support
    - Lastlog support
    - Memcache support
    - ncurses support
    - NLS support
    + OpenSSL support (OpenSSL 1.1.1g  21 Apr 2020)
    + PCRE support
    + POSIX ACL support
    - Redis support
    + Sendfile support
    + Shadow file support
    - Sodium support
    + Trace support
    + xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 4096
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10
```

TLS configuration:
```
# custom configuration
TLSEngine                 on
TLSProtocol               TLSv1.2 TLSv1.3
TLSOptions                NoSessionReuseRequired
TLSRequired               on
TLSCipherSuite            EECDH+AES128:EECDH+AES256:CHACHA20:!aPSK:!PSK:!DHE
TLSServerCipherPreference on

TLSRSACertificateFile     /etc/letsencrypt/live/ftp.xxx.eu/cert.pem
TLSRSACertificateKeyFile  /etc/letsencrypt/live/ftp.xxx.eu/privkey.pem
TLSCertificateChainFile   /etc/letsencrypt/live/ftp.xxx.eu/chain.pem
TLSDHParamFile            /etc/proftpd/dhparam2048.pem

#TLSSessionTickets         off
#TLSStapling               off
TLSSessionTickets         on
TLSStapling               on
TLSSessionCache           shm:/file=/run/proftpd/sess_cache
TLSStaplingCache          shm:/file=/run/proftpd/ocsp_pcache
```

### How to test / reproduce

I've tested with [sslyze](https://github.com/nabla-c0d3/sslyze/releases/):
```
john@DESKTOP-SI79RLD MINGW64 /c/devel/sslyze-3.0.6-exe
$ sslyze.exe --reneg --tlsv1_2 --starttls=ftp ftp.xxx.eu:21

 CHECKING HOST(S) AVAILABILITY
 -----------------------------

   ftp.xxx.eu:21                        => 88.198.xx.xx




 SCAN RESULTS FOR FTP.XXX.EU:21 - 88.198.xx.xx
 ------------------------------------------------

 * Session Renegotiation:
       Client-initiated Renegotiation:    VULNERABLE - Server honors client-initiated renegotiations
       Secure Renegotiation:              OK - Supported

 * TLS 1.2 Cipher suites:
     Attempted to connect using 158 cipher suites.

     The server accepted the following 7 cipher suites:
        TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256       256       ECDH: x25519 (253 bits)
        TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384             256       ECDH: prime256v1 (256 bits)
        TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384             256       ECDH: prime256v1 (256 bits)
        TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA                256       ECDH: prime256v1 (256 bits)
        TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256             128       ECDH: prime256v1 (256 bits)
        TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256             128       ECDH: prime256v1 (256 bits)
        TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA                128       ECDH: prime256v1 (256 bits)

     The group of cipher suites supported by the server has the following properties:
       Forward Secrecy                    OK - Supported
       Legacy RC4 Algorithm               OK - Not Supported

     The server has no preferred cipher suite.



 SCAN COMPLETED IN 8.84 S
 ------------------------

```

Thanks,
Simon

**Comment by Castaglia** [comment]
In attempting to reproduce this locally, I run `sslyze`, but see this:
```
# sslyze --reneg --tlsv1_2 --starttls=ftp 127.0.0.1:2121
...
     The server is configured to prefer the following cipher suite:
        TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256             128       ECDH: prime256v1 (256 bits)
``` 
So far as I can tell, the [`TLSServerCipherPreference`](http://www.proftpd.org/docs/contrib/mod_tls.html#TLSServerCipherPreference) directive is working properly.

For the topic of client-initiated renegotations:
```
 * Session Renegotiation:
       Client-initiated Renegotiation:    VULNERABLE - Server honors client-initiated renegotiations
       Secure Renegotiation:              OK - Supported
```
It is arguable just how big of a "vulnerability" this is.  There _are_ legitimate reasons a client might want to renegotiate a long-lived TLS session, as when transferring GB of data and wanting to refresh the session keys during that long transfer.

If the desire is to be able to disable such support for renegotiations, then I think I would need to patch the handling of the [`TLSRenegotiate`](http://www.proftpd.org/docs/contrib/mod_tls.html#TLSRenegotiate) directive, such that turning it off completely:
```
TLSRenegotiate off
```
would then disable support for _all_ renegotiations in OpenSSL.  Unfortunately OpenSSL does not have a way to allow server-initiated renegotiations, but *not* client-initiated renegotiations, that I could easily find.

**Comment by bratkartoffel** [comment]
Thanks for looking into this.

For the client initiated renegotions I found this: https://github.com/nodejs/node-v0.x-archive/issues/2726#issuecomment-3901047
This issue has a broken link to the connections.c, take a look at here: https://github.com/lighttpd/lighttpd1.4/blob/lighttpd-1.4.30/src/connections.c#L226

I think this would prevent malicious clients from consuming too much resources with renegotiations on the server, but also enabling it for the server for the reasons you mentioned.

For the preferred cipher suites I have to do further testing, it may be an issue with alpine / musl / specific openssl version?

Thanks,
Simon

**Comment by bratkartoffel** [comment]
It looks that it's an issue with sslyze itself. When enabling CHACHA20 Cipher you also set the SSL_OP_PRIORITIZE_CHACHA option. As this flag interferes with the default cipher ordering and negotiated cipher it seems that the tool cannot handle this situation correctly.
After some manual testing the behaviour of proftpd seems fine too me,.

**Comment by Castaglia** [comment]
Hmm.  Good point about the `SSL_OP_PRIORITIZE_CHACHA` option.  Based on this feedback, I may remove the setting of that option, precisely because it violates the principle of least surprise (_i.e._ "why isn't ProFTPD honoring my cipher preferences?").

**Comment by Castaglia** [comment]
With https://github.com/proftpd/proftpd/pull/1103 applied, the client-initiated renegotiations should be properly rejected.

Note that I think https://github.com/proftpd/proftpd/pull/1098 is still a nice refinement, to be merged in afterward (and only in master, not to be backported to the 1.3.7 branch).

**Comment by Castaglia** [comment]
This should now be fixed in master (and backported to the 1.3.7 branch).  Thanks!
